### PR TITLE
New E2E proposal: Validate that all the pods are removed when the job is deleted

### PR DESCRIPTION
**Pull request by globervinodhn**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
/kind feature
> /kind flake

**What this PR does / why we need it**: This PR requests for inclusion of a new E2E test proposal with respect to Jobs. This test verifies the behaviour between jobs and it's associated pods. When the owner entity i.e., the JOB is deleted, all the pods referenced by that JOB should also be deleted.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:
1. The PR verifies that the pods associated with the Job is deleted once the JOB is deleted.
2. The test takes around 42 to 48 seconds for execution (Approximately)
3. The test is found to be non-flaky.
4. This test can be later promoted to conformance.

**Does this PR introduce a user-facing change?**: NONE
release-note-none

cc @mgdevstack  @brahmaroutu 

/area conformance

### CODE REVIEW

**Comment by SataQiu** [comment]
Thanks @globervinodhn 
/ok-to-test
/release-note-none
/priority important-soon
/cc @SataQiu

File: test/e2e/apps/job.go [hunk 1]
```diff
@@ -48,6 +48,34 @@ var _ = SIGDescribe("Job", func() {
...
+		job, err := framework.CreateJob(f.ClientSet, f.Namespace.Name, job)
+		Expect(err).NotTo(HaveOccurred(), "failed to create job in namespace: %s", f.Namespace.Name)
+
+		By("Ensure pods equal to paralellism count is attached to the job")
+		err = framework.WaitForAllJobPodsRunning(f.ClientSet, f.Namespace.Name, job.Name, parallelism)
+		Expect(err).NotTo(HaveOccurred(), "failed to ensure number of pods associated with job is equal to parallelism count in namespace: %s", f.Namespace.Name)
+
+		running, err := framework.CheckForAllJobPodsRunning(f.ClientSet, f.Namespace.Name, job.Name, 2)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(running).To(BeTrue())
```
[the remark this answers is not in the document]
Comment (globervinodhn) [reply test/e2e/apps/job.go]: Thanks for the review. Descriptions provided for the two Expect calls.

File: test/e2e/apps/job.go [hunk 2]
```diff
@@ -48,6 +48,34 @@ var _ = SIGDescribe("Job", func() {
 		Expect(err).NotTo(HaveOccurred(), "failed to ensure job completion in namespace: %s", f.Namespace.Name)
 	})
 
+	/*
+		Testcase: Ensure that the jobs associated with the job are removed once the job is deleted
```
Comment (SataQiu) [inline test/e2e/apps/job.go]:
```suggestion
		Testcase: Ensure that the pods associated with the job are removed once the job is deleted
```

File: test/e2e/apps/job.go [hunk 3]
```diff
@@ -48,6 +48,34 @@ var _ = SIGDescribe("Job", func() {
...
+		Description: Create a job and ensure the associated pod count is equal to paralellism count. Delete the job and
+						ensure if the pods associated with the jobs have been removed
```
Comment (SataQiu) [inline test/e2e/apps/job.go]:
```suggestion
						ensure if the pods associated with the job have been removed
```

Comment (SataQiu) [reply test/e2e/apps/job.go]: added some minor comments.

File: hack/.shellcheck_failures [hunk 4]
```diff
@@ -27,6 +27,11 @@
 ./hack/cherry_pick_pull.sh
 ./hack/ginkgo-e2e.sh
 ./hack/grab-profiles.sh
+<<<<<<< HEAD
+=======
+./hack/jenkins/test-dockerized.sh
+./hack/jenkins/upload-to-gcs.sh
+>>>>>>> 77edb49262d13be2dff47e7fee9769001d421dd9
```
Comment (SataQiu) [inline hack/.shellcheck_failures]: You need to resolve these conflicts first!

**Comment by SataQiu** [comment]
ping @neolit123

File: cmd/kubeadm/app/phases/copycerts/copycerts.go [hunk 5]
```diff
@@ -89,7 +89,7 @@ func UploadCerts(client clientset.Interface, cfg *kubeadmapi.InitConfiguration,
 	fmt.Printf("[upload-certs] Storing the certificates in Secret %q in the %q Namespace\n", kubeadmconstants.KubeadmCertsSecret, metav1.NamespaceSystem)
 	decodedKey, err := hex.DecodeString(key)
 	if err != nil {
-		return err
+		return errors.Wrap(err, "error decoding certificate key")
```
Comment (neolit123) [inline cmd/kubeadm/app/phases/copycerts/copycerts.go]: this change seems out of scope for this PR title?

File: test/e2e/apps/job.go [hunk 6]
```diff
@@ -57,6 +57,30 @@ var _ = SIGDescribe("Job", func() {
...
+	/*
+		Testcase: Ensure that the pods associated with the job are removed once the job is deleted
+		Description: Create a job and ensure the associated pod count is equal to paralellism count. Delete the job and
+						ensure if the pods associated with the job have been removed
+	*/
+	ginkgo.It("should remove pods when job is deleted", func() {
+		ginkgo.By("Creating a job")
+		job := jobutil.NewTestJob("notTerminate", "all-pods-removed", v1.RestartPolicyNever, parallelism, completions, nil, backoffLimit)
+		job, err := jobutil.CreateJob(f.ClientSet, f.Namespace.Name, job)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "failed to create job in namespace: %s", f.Namespace.Name)
```
Comment (neolit123) [inline test/e2e/apps/job.go]: you this instead:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/util.go#L2063

**Review by neolit123 — COMMENTED** [review]
/sig apps

File: cmd/kubeadm/app/phases/copycerts/copycerts.go (hunk 5 above)
Comment (oomichi) [reply cmd/kubeadm/app/phases/copycerts/copycerts.go]: +1 for @neolit123 here

File: test/e2e/apps/job.go (hunk 6 above)
Comment (oomichi) [reply test/e2e/apps/job.go]: Yes, right.
Now `hack/verify-code-test.sh` enforces to use the function instead of `gomega.Expect(err).NotTo(gomega.HaveOccurred())`, so we need to replace here.

File: test/e2e/apps/job.go [hunk 7]
```diff
@@ -57,6 +57,30 @@ var _ = SIGDescribe("Job", func() {
...
+
+		ginkgo.By("Ensure pods equal to paralellism count is attached to the job")
+		err = jobutil.WaitForAllJobPodsRunning(f.ClientSet, f.Namespace.Name, job.Name, parallelism)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "failed to ensure number of pods associated with job %s is equal to parallelism count in namespace: %s", job.Name, f.Namespace.Name)
```
Comment (oomichi) [inline test/e2e/apps/job.go]: ditto

File: test/e2e/apps/job.go [hunk 8]
```diff
@@ -57,6 +57,30 @@ var _ = SIGDescribe("Job", func() {
...
+
+		ginkgo.By("Delete the job")
+		err = framework.DeleteResourceAndWaitForGC(f.ClientSet, batchinternal.Kind("Job"), f.Namespace.Name, job.Name)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "failed to delete the job in namespace: %s", f.Namespace.Name)
```
Comment (oomichi) [inline test/e2e/apps/job.go]: ditto

File: test/e2e/apps/job.go [hunk 9]
```diff
@@ -57,6 +57,30 @@ var _ = SIGDescribe("Job", func() {
...
+
+		ginkgo.By("Ensure the pods associated with the job are also deleted")
+		err = jobutil.WaitForAllJobPodsGone(f.ClientSet, f.Namespace.Name, job.Name)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred(), "failed to get PodList for job %s in namespace: %s", job.Name, f.Namespace.Name)
```
Comment (oomichi) [inline test/e2e/apps/job.go]: ditto

**Review by oomichi — COMMENTED** [review]
The test itself seems good for me. Few comments for passing the CI job.

/retest

**Comment by globervinodhn** [comment]
Closed by mistake

File: cmd/kubeadm/app/phases/copycerts/copycerts.go (hunk 5 above)
Comment (globervinodhn) [reply cmd/kubeadm/app/phases/copycerts/copycerts.go]: Thanks for the review. I have made the necessary changes.

**Comment by oomichi** [comment]
Thanks for updating.

/lgtm

**Review by timothysc — APPROVED** [review]
/assign @timothysc 
/lgtm 
/approve
